classes/local/breakoutexporter.php

```diff
@@ -0,0 +1,124 @@
+<?php
+
+
+namespace mod_zoom\local;
+use csv_export_writer;
+
+class breakoutexporter
+{
+    private $zoom;
+    private $course;
+    private $cm;
+    private $grouping;
+    private $groups;
+    private $groupmembers;
+    /**
+     * @var list of users that we've already seen, and the groups they're in.
+     */
+    private $seenbefore;
+
+    /**
+     * @var bool|int False if the data has never been loaded or a timestamp for when it was updated last.
+     */
+    private $dataloaded = false;
+
+    public function __construct(StdClass $zoom)
+    {
+        $this->zoom = zoom;
+        list($course, $cm) = get_course_and_cm_from_instance($zoom, 'zoom');
+        $this->course = $course;
+        $this->cm = $cm;
+    }
+
+    /**
+     * Get All the data we need for export.
+     */
+    public function filldata() {
+        // Grouping details.
+        $this->grouping = groups_get_grouping($this->zoom->breakoutgrouping);
+
+        // List of groups in the grouping
+        $this->groups = groups_get_all_groups($this->course->id, $this->grouping->id);
+
+        if (!empty($this->groups)) {
+            $members = [];
+            foreach($this->groups as $group) {
+                $gms = groups_get_members($group->id);
+                $members[$group->id] = [];
+                foreach($gms as $user) {
+                    if (!isset($this->seenbefore[$user->id])) {
+                        $this->seenbefore[$user->id] = [$group->id];
+                        $members[$group->id][] = $user;
+                    } else {
+                        $this->seenbefore[$user->id][] = $group->id;
+                    }
+                }
+            }
+            $this->groupmembers = $members;
```

Comment:
Here's my thought on this section, keep the parts you like.
```suggestion
        $this->groupmembers = [];
        if (!empty($this->groups)) {
            $seenbefore = [];
            foreach ($this->groups as $group) {
                $gms = groups_get_members($group->id);
                foreach ($gms as $user) {
                    // A users can only be in one breakout group, so use the first one.
                    if (!isset($seenbefore[$user->id])) {
                        $this->groupmembers[$group->id][$user->id] = $user;
                        $seenbefore[$user->id] = true;
                    }
                }
            }
```

Reply:
I did this this way round so that the $seenbefore array would have enough data to be able to explain to the end user why a user may be appearing in an unexpected group (e.g. if they expect them in a later group).

This would allow the generation of something like;
Michael Hughes appeared in multiple groups (Group A*, Group B, Group Z). They will be in breakout room "Group A").

Reply:
I understand, but there isn't a function that within the class that uses it, so since it is private, no one can access the value. If you want to keep it, please initialize it to empty when you call `filldata()` so that multiple calls to `filldata()` return the same value.

Reply:
You're right, not yet there isn't! But I'll bear this in mind when I'm looking at this again (hopefully this week).